code/common/pubsub.q

```diff
@@ -1,14 +1,156 @@
+ };
+
+// Add handle for subscriber using old API (filter is list of syms)
+selfiltered:{[x;y]
+  filts:enlist (in;`sym;enlist y);
+  if[`error~first test:.[?;(.stpps.schemas[x];filts;0b;());{(`error;x)}];
+    .lg.e[`addfiltered;"Invalid query parameters provided: ",last test];
+    :test
+   ];
+  `.stpps.subrequestfiltered upsert (x;.z.w;filts;());
+ };
+
+// Publish table data
+pub:{[t;x]
+  if[count x;
+    if[count h:subrequestall[t];-25!(h;(`upd;t;x))];
+    if[t in subrequestfiltered`tbl;
+      {[t;x]data:?[t;x`filts;0b;x`columns];neg[x`handle](`upd;t;data)}
+      [t;]each select handle,filts,columns from subrequestfiltered where tbl=t
+    ];
+  ];
+ };
+
+// publish and clear tables
+pubclear:{
+ .stpps.pub'[x;value each x,:()];
+ @[`.;x;:;.stpps.schemasnoattributes[x]];
+ }
+
+// Remove handle from subrequestall
+delhandle:{[t;h]
+  @[`.stpps.subrequestall;t;except;h];
+ };
+
+// Remove handle from subrequestfiltered
+delhandlef:{[t;h]
+  delete from  `.stpps.subrequestfiltered where tbl=t,handle=h;
+ };
+
+// Remove all handles when connection closed
+closesub:{[h]
+  delhandle[;h]each t;
+  delhandlef[;h]each t;
+ };
+
+// Set up table and schema information
+init:{
+  // Grab non-keyed tables from root namespace and store their attributes
+  .stpps.t:(tables[] where 98h=type each value each tables[]) except `currlog;
+  .stpps.schemas:.stpps.t!value each .stpps.t;
+
+  // Strip attributes from tables and store new schemas if process is an STP and a dictionary of table column names
+  if[`segmentedtickerplant~.proc.proctype;
+    {@[x;cols x;`#]} each .stpps.t;
+    .stpps.schemasnoattributes:.stpps.t!value each .stpps.t
+    ];  
+  .stpps.tabcols:.stpps.t!cols each .stpps.t;
+ };
+
+\d .
+
+// Call closesub function after initial .z.pc call on disconnect
+.z.pc:{[f;x] @[f;x;()];.stpps.closesub x} @[value;`.z.pc;{{}}];
+
+// Function called on subscription
+// Subscriber will call with null y parameter in sub all mode
+// In sub filtered mode, y will contain tables to subscribe to and filters to apply
+.u.sub:{[x;y]
+  if[x~`;:.u.sub[;y] each .stpps.t];
+  if[not x in .stpps.t;
+    .lg.e[`sub;m:"Table ",string[x]," not in list of stp pub/sub tables"];
+    :(x;m)
+  ];
+  $[y~`;.stpps.suball[x];.stpps.subfiltered[x;y]]
+ };
+
+// Define .ps wrapper functions
+.ps.loaded:1b;
+.ps.publish:.stpps.pub;
+.ps.subscribe:.u.sub;
+.ps.init:.stpps.init;
+.ps.initialise:{.ps.init[];.ps.initialised:1b};
+
+// Allow a non-kdb+ subscriber to subscribe with strings for simple conditions - return string to subscriber
+.ps.subtable:{[tab;syms]
+  .lg.o[`subtable;"Received a simple string subscription request."];
+  val:.[.u.sub;(`$tab;$[count syms;::;first] `$csv vs syms);{:(`error;"Error: ",x)}];
+  $[`error~first val;last val;"Subscription successful!"]
+ };
+
+// Allow a non-kdb+ subscriber to subscribe with strings for complex conditions - return string to subscriber
+.ps.subtablefiltered:{[tab;filters;columns]
```

Comment:
Need to make sure all these external interface functions are documented
(I haven't checked)

Reply:
They are